cli/azd/design-notes/environments.md

```diff
@@ -0,0 +1,271 @@
+5. An ARM Deployment is created to deploy the bicep template with the `main.parameters.json` file from step 4. The name of this deployment matches the environment name and the target subscription is pulled from the `AZURE_SUBSCRIPTION_ID` environment variable.
+6. As described earlier, when the deployment completes, all outputs are merged in to the `.env` file for the environment.
+
+By convention, our templates also tag resources with `azd-env-name` and there has been discussion about `azd` using that tag when running queries to discover resources. We've done something similar with `azd-service-name` and that pattern has worked out well for us and plays nicely in the infrastructure as code world.
+
+This system allows you to create a multiple environments, each with a different name, and provision a unique copy of the infrastructure for the application.
+
+## Current Challenges
+
+The major issue we face with the current design is that we've introduced a new concept, an environment, and it is on the critical path of every `azd` invocation. We need to ensure this concept is simple to understand and worthwhile and clearly articulate how it relates to the the rest of the system (and ideally existing Azure concepts). It's not always clear how settings in the environment impact infrastructure or the rest of `azd` and what's based on convention and what is required by the tooling. There are also "tiers" of configuration values now - `AZURE_ENV_NAME`, `AZURE_LOCATION`, `AZURE_PRINCIPAL_ID` and `AZURE_SUBSCRIPTION_ID` are very different from other values. We also have a tension on how to describe what an environment name does. Our templates today use it to influence the names of some resources (usually the root resource group that is created) but that's simply a convention we've adopted and so we are wary of adding that as "help text" for an environment name.
+
+The value of environments doesn't pay off until you want to start creating multiple copies of the infrastructure that backs your application, but the complexity cost is incurred as a soon as a user walks up to `azd`.
+
+I think that with a few small tweaks to our model we can make it much easier to understand and describe - while at the same time allowing us to defer introduction of the concept until a user actually cares about the value an environment provides.
+
+## Proposed Model
+
+### Overview
+
+As before, we still have the concept of an environment in our system.  All `azd` commands still accept an `-e` parameter to select what environment is used, and `azd env select` can be used to select the default environment. However, if no environment is selected, `azd` creates a new environment with the name `default`. In addition to a name every environment gets a unique id (a v4 UUID) assigned to it. These values are recorded in `.azure/<environment-name>/config.json`:
```

Comment:
I'm concerned about creating an environment with a name that the user didn't select. I think that is further abstracting the concept of environment rather than embracing it. 

Instead we should guide them through the process, explain what things means, and help them make informed decisions.